internal/resources/metal/vlan/models.go

```diff
@@ -0,0 +1,98 @@
+	ProjectID          types.String `tfsdk:"project_id"`
+	VlanID             types.String `tfsdk:"vlan_id"`
+	Vxlan              types.Int64  `tfsdk:"vxlan"`
+	Facility           types.String `tfsdk:"facility"`
+	Metro              types.String `tfsdk:"metro"`
+	Description        types.String `tfsdk:"description"`
+	AssignedDevicesIds types.List   `tfsdk:"assigned_devices_ids"`
+}
+
+func (m *DataSourceModel) parse(vlan *packngo.VirtualNetwork) diag.Diagnostics {
+	m.ID = types.StringValue(vlan.ID)
+
+	if vlan.Project.ID != "" {
+		m.ProjectID = types.StringValue(vlan.Project.ID)
+	}
+
+	m.VlanID = types.StringValue(vlan.ID)
+	m.Description = types.StringValue(vlan.Description)
+	m.Vxlan = types.Int64Value(int64(vlan.VXLAN))
+
+	var metroCode, facilityCode types.String
+	if vlan.Facility != nil {
+		facilityCode = types.StringValue(vlan.Facility.Code)
+		metroCode = types.StringValue(strings.ToLower(vlan.Facility.Metro.Code))
+	}
+	// version of this resource. StateFunc doesn't exist in terraform and it requires implementation
+	// of bespoke logic before storing state. To ensure backward compatibility we ignore lower/upper
+	// case diff for now, but we may want to require input upper case
+	if !strings.EqualFold(m.Facility.ValueString(), facilityCode.ValueString()) {
+		m.Facility = facilityCode
+	}
+
+	if vlan.Metro != nil {
+		metroCode = types.StringValue(strings.ToLower(vlan.Metro.Code))
+	}
+	if !strings.EqualFold(m.Metro.ValueString(), metroCode.ValueString()) {
+		m.Metro = metroCode
+	}
+
+	deviceIds := make([]types.String, 0, len(vlan.Instances))
+	for _, device := range vlan.Instances {
+		deviceIds = append(deviceIds, types.StringValue(device.ID))
+	}
+
+	return m.AssignedDevicesIds.ElementsAs(context.Background(), &deviceIds, false)
+}
+
+type ResourceModel struct {
+	ID          types.String `tfsdk:"id"`
+	ProjectID   types.String `tfsdk:"project_id"`
+	Vxlan       types.Int64  `tfsdk:"vxlan"`
+	Facility    types.String `tfsdk:"facility"`
+	Metro       types.String `tfsdk:"metro"`
+	Description types.String `tfsdk:"description"`
+}
+
+func (m *ResourceModel) parse(vlan *packngo.VirtualNetwork) diag.Diagnostics {
+	m.ID = types.StringValue(vlan.ID)
+	m.Description = types.StringValue(vlan.Description)
+	m.Vxlan = types.Int64Value(int64(vlan.VXLAN))
+
+	if vlan.Project.ID != "" {
+		m.ProjectID = types.StringValue(vlan.Project.ID)
+	}
+
+	var metroCode types.String
+	facilityCode := types.StringNull()
+	if vlan.Facility != nil {
+		facilityCode = types.StringValue(vlan.Facility.Code)
+		metroCode = types.StringValue(strings.ToLower(vlan.Facility.Metro.Code))
+	}
+	// version of this resource. StateFunc doesn't exist in terraform and it requires implementation
+	// of bespoke logic before storing state. To ensure backward compatibility we ignore lower/upper
+	// case diff for now, but we may want to require input upper case
+	if !strings.EqualFold(m.Facility.ValueString(), facilityCode.ValueString()) {
+		m.Facility = facilityCode
```

Comment:
I see a bunch of failures in CI due to facility changing from `""` to nil after apply.  Not sure where it could be coming from, but maybe here?  This should probably just set `m.Facility = facilityCode` without the `EqualFold` check so that `m.Facility` is always set.

Reply:
Even with the recommended changes, there are still a large number of tests failing due to facility being nil after apply.  Are all of the VLAN tests using a dedicated connection or do some of them use shared connections so they can be run locally?

Reply:
All the tests are passing now. Although a few of them failing due to device unavailablility and timeout